test/test_fork.py

```diff
@@ -0,0 +1,147 @@
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+# http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+"""Test that pymongo is fork safe."""
+
+import os
+import threading
+from multiprocessing import Pipe
+from test import IntegrationTest, client_context
+from typing import Any, Callable
+from unittest import skipIf
+from unittest.mock import patch
+
+from bson.objectid import ObjectId
+from pymongo import MongoClient
+from pymongo.lock import _ForkLock
+
+
+@client_context.require_connection
+def setUpModule():
+    pass
+
+
+class ForkThread(threading.Thread):
+    def __init__(self, exit_cond: Callable[[], int] = lambda: 0):
+        super().__init__()
+        self.pid: int = -1  # Indicate we haven't started.
+        self.exit_cond = exit_cond
+
+    def run(self):
+        if self.pid < 0:
+            self.pid = os.fork()
+            if self.pid == 0:
+                os._exit(self.exit_cond())
+
+
+class LockWrapper:
+    def __init__(self, fork_thread: ForkThread, lock_type: Any = _ForkLock):
+        self.__lock = lock_type()
+        self.fork_thread = fork_thread
+
+    def __enter__(self):
+        self.__lock.__enter__()
+        if not self.fork_thread.is_alive():
+            self.fork_thread.start()
+
+    def __exit__(self, exc_type, exc_value, traceback):
+        self.__lock.__exit__(exc_type, exc_value, traceback)
+
+    def __getattr__(self, item):
+        return getattr(self.__lock, item)
+
+
+# Not available for versions of Python without "register_at_fork"
+@skipIf(
+    not hasattr(os, "register_at_fork"), "register_at_fork not available in this version of Python"
+)
+class TestFork(IntegrationTest):
+    def setUp(self):
+        self.db = self.client.pymongo_test
+        self.fork_thread = ForkThread()
+
+    def test_lock_client(self):
+        """
+        Forks the client with some items locked.
+        Parent => All locks should be as before the fork.
+        Child => All locks should be reset.
+        """
+
+        def exit_cond():
+            # Checking all the locks after forking this way is highly
+            # probabilistic as more locking behavior may ensue. Instead we
+            # check that we can acquire _insertion_lock, meaning we have
+            # successfully completed.
+            with _ForkLock._insertion_lock:
+                return 0  # success
+
+        self.fork_thread.exit_cond = exit_cond
+        with patch.object(
+            self.db.client, "_MongoClient__lock", LockWrapper(fork_thread=self.fork_thread)
+        ):
```

Comment:
I had it doing that a few commits ago, and I realized it's probably more realistic to fork from another thread during a critical region. I think wrapping it also allows us to test lock behavior in functions where they are used.

Reply:
I still think it would be preferable to remove LockWrapper or at least simplify it. For example LockWrapper should not be creating a new lock. Instead it should wrap the existing lock we're trying to test. IE:
```
        with patch.object(
            self.db.client, "_MongoClient__lock", LockWrapper(fork_thread=fork_thread, lock=self.client._MongoClient__lock)
        ):

```

Second, there is still a race where the ForkThread might not even run until _after_ the lock has been released.


Manually acquiring the lock is a much simpler and provides the same test coverage so I vote we remove ForkThread and LockWrapper altogether in favor of the approach described above:
```python
with self.client._MongoClient__lock:
    # call fork() assert child process can run an operation on the client.

with ObjectId._inc_lock:
    # call fork() assert child process can still create an ObjectId.
```